### ISSUE TITLE
Support for ARM32 in NativeAOT

### ISSUE THREAD

**Issue by bjornen77** [dotnet/runtimelab#833]
Is there any plans to support ARM32? Basically so that you from a PC running Linux(amd64) could do:
dotnet publish -r linux-arm -c release

Or could you already today compile applications for ARM32 using NativeAOT?

**Comment by jkotas** [comment]
The core team working on the NativeAOT experiment in this repo does not have plans for it, but we are happy to accept contributions to make it work.

**Comment by josephmoresena** [comment]
I recall CoreRT had support on ARM but I never used it. I was actually looking for that cause I want to experiment with JNI libs but using Native NOT.

**Comment by jkotas** [comment]
Yes, we have received some contributions for ARM support in the past but it did not get pushed far enough to get official builds enabled, etc.

**Comment by michaldobrodenka** [comment]
Would be great if at least this could be the way for NET to support ARMv6. Compile to llvm -> ARMv6 or ARMv7 native.

**Comment by trampster** [comment]
Arm is potentially a better use case for NativeAOT than x86. Arm devices tend to be used in resource constrained environments like IOT devices, raspberry pi's, etc.  And can benefit more from the advantages of NativeAOT.

The team I work on has an embedded linux device using OpenEmbedded as the operating system on an arm32 chip. We had to reject .NET as an option because we simply do not have the space for it, and we need fast startup time.

NativeAOT seems like the perfect solution but we can't use it because it doesn't support arm32.

**Comment by kant2002** [comment]
Right now building of NativeAOT on Raspberry 3 (I assume that this is ARM32) fails with following error

```
  CMake Error at nativeaot/Runtime/Full/CMakeLists.txt:25 (add_library):
    Cannot find source file:

      /home/pi/runtimelab/src/coreclr/nativeaot/Runtime/arm/GcProbe.S

    Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
    .hpp .hxx .in .txx


  CMake Error at nativeaot/Runtime/Full/CMakeLists.txt:25 (add_library):
    No SOURCES given to target: Runtime.ServerGC


  CMake Error at nativeaot/Runtime/Full/CMakeLists.txt:23 (add_library):
    No SOURCES given to target: Runtime


  CMake Generate step failed.  Build files cannot be regenerated correctly.
  ~/runtimelab/src/coreclr
```

**Comment by ChrML** [comment]
Anyone got an overview of what's left to get NativeAOT running on ARM32?

**Comment by kant2002** [comment]
Depending on you ability to take risks.

NativeAOT in basic form runs on ARM32 now. I mean Hello World running. Downside is that you need build it yourself. That's not compilcated just time consuming.

If you want do not think too much about building NativeAOT, then following items left:
- [x] Add ARM to CI build. Need general approval from @jkotas . Something like this https://github.com/dotnet/runtimelab/pull/1509
- [ ] Teach tools that ARM is supported. Something like that https://github.com/dotnet/runtimelab/pull/1387
- [x] For purity (not affects anybody). Wait while https://github.com/dotnet/runtime/pull/58597 find it's way to this repo.
- [ ] Run tests at least once on ARM to collect data on what's broken.

**Comment by jkotas** [comment]
> Add ARM to CI build. Need general approval from @jkotas

I would be fine with adding Linux ARM to the CI build.

**Comment by pver** [comment]
#1509 seems to be (partially) undone by https://github.com/dotnet/runtimelab/commit/562aea2cb7d449d6e2e697df8cac56b599ec564d ?

**Comment by kant2002** [comment]
@pver you can still build ARM32 packages. Support is still somewhat limited and do not tested enough. Can you tell more about your case? Curious why do you need ARM32

**Comment by pver** [comment]
@kant2002 currently have a .net application with mono on an armv7l platform, but running on some memory/performance issues/constraints. So I was test driving NativeAOT to see if that could improve things in the future :) 
Tried running with the arm docker image and publish for linux-arm, but kept running against 
`/root/.nuget/packages/microsoft.dotnet.ilcompiler/7.0.0-alpha.1.21455.2/build/Microsoft.NETCore.Native.Publish.targets(71,5): error : Add a PackageReference for 'runtime.linux-x64.Microsoft.DotNet.ILCompiler' to allow cross-compilation for arm [/test/HelloCoreRTWorld.csproj]
       Done executing task "Error" -- FAILED`
even though targetting **linux-arm** and including runtime.linux-arm.Microsoft.DotNet.ILCompiler (latest available for that package is 7.0.0-alpha.1.21455.2)

But I'll do some further tests with it, building the packages myself as you suggested and see how far I can get with those

**Comment by michaldobrodenka** [comment]
@kant2002 I'm also using dotnet mainly on ARM devices, for IoT. Currently publishing using self-contained, since I don't have much space left on the device for full framework. If NativeAOT would give me smaller package or or faster startup time, it would be great. X86/64 devices have most of the times more storage, RAM or CPU power so there it's not so important I think.

**Comment by kant2002** [comment]
> Tried running with the arm docker image and publish for linux-arm

@pver are you run docker on ARM hardware? It seems to be you are running Docker image on Linux x64, and as such you may have to follow instructions for cross-compilation. https://github.com/dotnet/runtimelab/blob/feature/NativeAOT/docs/using-nativeaot/compiling.md#cross-architecture-compilation

@pver and @michaldobrodenka if you discover any specific issues, would be good to post it in this repo, so they can be tracked. My experience with this repo, is that all issues taken seriously, it maybe just take a bit of time to solve all of them.

**Comment by pver** [comment]
@kant2002 thanks for the hint! I'm indeed running Docker from a Linux x64. I tried to build with the mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-16.04-cross-20200413125008-09ec757 (and the newer ubuntu-16.04-cross-20210719121212-8a8d3be) docker image for arm32, but maybe that's not enough to fully cross compile. I'll try with the manual steps to see if that runs better and create a new ticket if I keep running into issues.

**Comment by TheGuy920** [comment]
Has this still not been resolved? It's been nearly 2 years 😢 
`Unable to find package runtime.linux-arm.Microsoft.DotNet.ILCompiler. No packages exist with this id in source(s): C:\Program Files\dotnet\library-packs, Microsoft Visual Studio Offline Packages, nuget.org	*	*.csproj	1`

**Comment by pver** [comment]
Very strange indeed, it seems to be supported based on this:
https://github.com/dotnet/runtime/blob/main/src/installer/pkg/projects/Microsoft.DotNet.ILCompiler/ILCompilerRIDs.props

But it's just not published? (only published for x64 platform packages at the moment). Or is this approach already in the spirit of https://github.com/dotnet/runtime/discussions/71042 ?

**Comment by kant2002** [comment]
As probably the last person who touch code for ARM32 I can say following.
ARM32 is community maintained. I was the community, and I lost interest in this. During the time when I work, I receive full support from core .NET contributors. I suppose that if somebody else would improve ARM32 in dotnet/runtime nobody would object too much. So far, there interest in consuming .NET on Arm32, but not contibute back. Technically ARM32 was somewhat working, but after two years I expect it bitrot. If somebody serious about Arm32 I think it should build dotnet/runtime locally, and try to compile application using NativeAOT using locally build bits.

Instructions to get started with NativeAOT contribution can be found here https://github.com/dotnet/runtime/blob/main/docs/workflow/building/coreclr/nativeaot.md

**Comment by jkotas** [comment]
> it seems to be supported based on this:
https://github.com/dotnet/runtime/blob/main/src/installer/pkg/projects/Microsoft.DotNet.ILCompiler/ILCompilerRIDs.props

This comment is misleading. I am going to delete it. This is list of RIDs that Microsoft.DotNet.ILCompiler may be pontetially built for, but it does not imply any kind of support.

@kant2002 summary is accurate. We will be happy to accept contributions for native AOT Arm32 port.

**Comment by pver** [comment]
> This comment is misleading. I am going to delete it

Wasn't aware about censorship, I can also remove that line myself if you want 😉 I never pretended to declare it as supported, I just pointed out there are traces of arm32 in there. With supported I actually meant the same as what you wrote, 'potentially be built for', but apparently isn't build and published at the moment. Your comment confirms this is on purpose, so thanks for the feedback.

> @kant2002 summary is accurate. We will be happy to accept contributions for native AOT Arm32 port

I was indeed going to have a look at that. If I would add contributions, will it then potentially be built? Or would it still be a 'build it yourself' feature?

**Comment by filipnavara** [comment]
Let's close this in favor of https://github.com/dotnet/runtime/issues/97729.

**Comment by guodawei2026** [comment]
@michaldobrodenka  http://icsharptech.com/  
.net10 +nativeaot+ arm mcu（arm32）
